### PR TITLE
[vectorized](udf) java udf support array type

### PR DESCRIPTION
**Pull request by zhangstar333**
# Proposed changes

> support java udf using array type, like:

` CREATE FUNCTION java_udf_array_return_int_test(array<int>) RETURNS array<int> PROPERTIES (
            "file"="file://${jarPath}",
            "symbol"="org.apache.doris.udf.ArrayReturnArrayIntTest",
            "type"="JAVA_UDF"
        );`
Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

* [ ] Does it affect the original behavior
* [ ] Has unit tests been added
* [ ] Has document been added or modified
* [ ] Does it need to update dependencies
* [ ] Is this PR support rollback (If NO, please explain WHY)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: regression-test/suites/javaudf_p0/test_javaudf_array_int.groovy [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+// specific language governing permissions and limitations
+// under the License.
+
+import org.codehaus.groovy.runtime.IOGroovyMethods
+
+import java.nio.charset.StandardCharsets
+import java.nio.file.Files
+import java.nio.file.Paths
+
+suite("test_javaudf_array_int") {
```
Comment (HappenLee) [inline regression-test/suites/javaudf_p0/test_javaudf_array_int.groovy:24]: rename `test_javaudf_array`

Comment (HappenLee) [inline regression-test/suites/javaudf_p0/test_javaudf_array_int.groovy:24]: rename `test_javaudf_array`

File: be/src/vec/functions/function_java_udf.cpp [hunk 2]
```diff
@@ -194,6 +228,74 @@ Status JavaFunctionCall::execute(FunctionContext* context, Block& block,
...
+        ColumnArray* array_col = assert_cast<ColumnArray*>(data_col.get());                        \
+        ColumnNullable& array_nested_nullable =                                                    \
+                assert_cast<ColumnNullable&>(array_col->get_data());                               \
+        auto data_column_null_map = array_nested_nullable.get_null_map_column_ptr();               \
+        auto data_column = array_nested_nullable.get_nested_column_ptr();                          \
+        auto& offset_column = array_col->get_offsets_column();                                     \
+        int increase_buffer_size = 0;                                                              \
+        int32_t buffer_size = JniUtil::IncreaseReservedBufferSize(increase_buffer_size);           \
+        offset_column.reserve(num_rows);                                                           \
+        offset_column.resize(num_rows);                                                            \
```
Comment (HappenLee) [inline be/src/vec/functions/function_java_udf.cpp:241]: only do resize is enough？

File: be/src/vec/functions/function_java_udf.cpp [hunk 3]
```diff
@@ -194,6 +228,74 @@ Status JavaFunctionCall::execute(FunctionContext* context, Block& block,
...
+        *(jni_ctx->output_offsets_ptr) =                                                           \
+                reinterpret_cast<int64_t>(offset_column.get_raw_data().data);                      \
+        data_column_null_map->reserve(buffer_size);                                                \
+        data_column_null_map->resize(buffer_size);                                                 \
```
Comment (HappenLee) [inline be/src/vec/functions/function_java_udf.cpp:245]: same to the up

File: be/src/vec/functions/function_java_udf.cpp [hunk 4]
```diff
@@ -194,6 +228,74 @@ Status JavaFunctionCall::execute(FunctionContext* context, Block& block,
...
+        jni_ctx->output_intermediate_state_ptr->buffer_size = buffer_size;                         \
+        if (data_column->is_column_string()) {                                                     \
+            ColumnString* str_col = assert_cast<ColumnString*>(data_column.get());                 \
+            ColumnString::Chars& chars = assert_cast<ColumnString::Chars&>(str_col->get_chars());  \
+            ColumnString::Offsets& offsets =                                                       \
+                    assert_cast<ColumnString::Offsets&>(str_col->get_offsets());                   \
+            chars.reserve(buffer_size);                                                            \
+            chars.resize(buffer_size);                                                             \
+            offsets.reserve(buffer_size);                                                          \
+            offsets.resize(buffer_size);                                                           \
```
Comment (HappenLee) [inline be/src/vec/functions/function_java_udf.cpp:259]: do not need reserve

File: fe/java-udf/src/main/java/org/apache/doris/udf/BaseExecutor.java [hunk 5]
```diff
@@ -206,25 +220,299 @@ protected Object[] allocateInputObjects(long row, int argClassOffset) throws Udf
...
+                        UdfUtils.getAddressAtOffset(inputOffsetsPtrs, argIdx)) + 8L * (row - 1)));
+        long offsetEnd = Integer.toUnsignedLong(UdfUtils.UNSAFE.getInt(null, UdfUtils.UNSAFE.getLong(null,
+                UdfUtils.getAddressAtOffset(inputOffsetsPtrs, argIdx)) + 8L * row));
+
+        switch (type.getPrimitiveType()) {
+            case BOOLEAN: {
+                ArrayList<Boolean> data = new ArrayList<>();
+                for (long offsetRow = offsetStart; offsetRow < offsetEnd; ++offsetRow) {
+                    if ((UdfUtils.UNSAFE.getByte(null, UdfUtils.UNSAFE.getLong(null,
+                            UdfUtils.getAddressAtOffset(inputArrayNullsPtrs, argIdx)) + offsetRow) == 1)) {
```
Comment (HappenLee) [inline fe/java-udf/src/main/java/org/apache/doris/udf/BaseExecutor.java:261]: should only call `UdfUtils.UNSAFE.getLong(null,
                            UdfUtils.getAddressAtOffset(inputArrayNullsPtrs, argIdx))` one time

File: fe/java-udf/src/main/java/org/apache/doris/udf/BaseExecutor.java [hunk 6]
```diff
@@ -206,25 +220,299 @@ protected Object[] allocateInputObjects(long row, int argClassOffset) throws Udf
...
+                ArrayList<BigInteger> data = new ArrayList<>();
+                for (long offsetRow = offsetStart; offsetRow < offsetEnd; ++offsetRow) {
+                    if ((UdfUtils.UNSAFE.getByte(null, UdfUtils.UNSAFE.getLong(null,
+                            UdfUtils.getAddressAtOffset(inputArrayNullsPtrs, argIdx)) + offsetRow) == 1)) {
+                        data.add(null);
+                    } else {
+                        long value = UdfUtils.UNSAFE.getLong(null,
+                                UdfUtils.UNSAFE.getLong(null, UdfUtils.getAddressAtOffset(inputBufferPtrs, argIdx))
+                                        + 16L * offsetRow);
+                        byte[] bytes = new byte[16];
```
Comment (HappenLee) [inline fe/java-udf/src/main/java/org/apache/doris/udf/BaseExecutor.java:446]: the new should out for loop

File: fe/java-udf/src/main/java/org/apache/doris/udf/BaseExecutor.java [hunk 7]
```diff
@@ -380,11 +668,435 @@ protected boolean storeUdfResult(Object obj, long row, Class retClass) throws Ud
...
+                    return false;
+                }
+                for (int i = 0; i < num; ++i) {
+                    Byte value = data.get(i);
+                    if (value == null) {
+                        UdfUtils.UNSAFE
+                                .putByte(UdfUtils.UNSAFE.getLong(null, outputArrayNullPtr) + (offset + i), (byte) 1);
+                    } else {
+                        UdfUtils.UNSAFE
+                                .putByte(UdfUtils.UNSAFE.getLong(null, outputArrayNullPtr) + (offset + i), (byte) 0);
```
Comment (HappenLee) [inline fe/java-udf/src/main/java/org/apache/doris/udf/BaseExecutor.java:722]: only call `UdfUtils.UNSAFE.getLong(null, outputArrayNullPtr)` one time

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.94 seconds
 stream load tsv:          480 seconds loaded 74807831229 Bytes, about 148 MB/s
 stream load json:         36 seconds loaded 2358488459 Bytes, about 62 MB/s
 stream load orc:          69 seconds loaded 1101869774 Bytes, about 15 MB/s
 stream load parquet:          29 seconds loaded 861443392 Bytes, about 28 MB/s
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20230217060348_clickbench_pr_98729.html

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**